sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs

```diff
@@ -206,86 +209,48 @@ private Request CreateAuthRequest(MsiType msiType, string[] scopes, string clien
-                // was user canceled we don't want to handle the exception so s_identityAvailable will
-                // remain unset, as we still haven't determined if the imds endpoint is available.
-                catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
-                {
-                    return false;
-                }
-                // in the case that the request failed in a manner that we didn't receive any response other,
-                // than being canceled, take this to indicate that the imds endpoint is not available.
-                catch (Exception e) when (!(e is OperationCanceledException))
-                {
-                    return false;
+                    var success = result.AsyncWaitHandle.WaitOne(ImdsAvailableTimeoutMs);
+
+                    return success && client.Connected;
                 }
             }
+            catch
+            {
+                return false;
+            }
         }
 
-        private async Task<bool> ImdsAvailableAsync(CancellationToken cancellationToken)
+        public virtual async Task<bool> ImdsAvailableAsync(CancellationToken cancellationToken)
         {
-            // send a request without the Metadata header.  This will result in a failed request,
-            // but we're just interested in if we get a response before the timeout of 500ms
-            // if we don't get a response we assume the imds endpoint is not available
-            using (Request request = _pipeline.HttpPipeline.CreateRequest())
+            // try to create a TCP connection to the IMDS IP address. If the connection can be established
+            // we assume that IMDS is available. If connecting times out or fails to connect assume that
+            // IMDS is not available in this environment.
+            try
             {
-                request.Method = RequestMethod.Get;
-
-                request.Uri.Reset(s_imdsEndpoint);
-
-                request.Uri.AppendQuery("api-version", ImdsApiVersion);
-
-                CancellationToken imdsTimeout = new CancellationTokenSource(ImdsAvailableTimeoutMs).Token;
-
-                try
+                using (var client = new TcpClient())
                 {
-                    AzureIdentityEventSource.Singleton.ProbeImdsEndpoint(request.Uri);
-
-                    Response response = await _pipeline.HttpPipeline.SendRequestAsync(request, CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, imdsTimeout).Token).ConfigureAwait(false);
+                    var result = client.BeginConnect(s_imdsHostIp, s_imdsPort, null, null);
 
-                    AzureIdentityEventSource.Singleton.ImdsEndpointFound(request.Uri);
+                    var success = await Task.Run<bool>(() => result.AsyncWaitHandle.WaitOne(ImdsAvailableTimeoutMs), cancellationToken).ConfigureAwait(false);
```

Comment:
There is a more efficient way of doing this https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/interop-with-other-asynchronous-patterns-and-types?redirectedfrom=MSDN#WHToTap but I'm not sure how much we care.

Reply:
Do we care about the optimization? The more efficient way complicates the code a bit.

Reply:
I agree. Since this should only be called once per process I think that we can leave the simplified code here and optimize later on if we find it to be a problem.